src/reader.rs

```diff
@@ -276,4 +349,211 @@ impl Reader {
+            imm_memtable: VecDeque::new(),
+            imm_wal: VecDeque::new(),
+        }
+    }
+
+    fn mock_level_getters<'a>(
+        row_entries: Vec<Option<RowEntry>>,
+    ) -> Vec<BoxFuture<'a, Result<Option<RowEntry>, SlateDBError>>> {
+        row_entries
+            .into_iter()
+            .map(|entry| async move { Ok(entry) }.boxed())
+            .collect()
+    }
+
+    struct LevelGetExpireTestCase {
+        entries: Vec<Option<RowEntry>>,
+        expected: Option<Bytes>,
+    }
+
+    #[tokio::test]
+    #[rstest]
+    #[case(LevelGetExpireTestCase {
+        entries: vec![
+            Some(RowEntry::new(
+                Bytes::from_static(b"key"),
+                ValueDeletable::Value(Bytes::from_static(b"v1")),
+                10,
+                Some(10000 - 2000),
+                Some(10000 - 1000),
+            )), // already expired, should be None
+            Some(RowEntry::new(
+                Bytes::from_static(b"key"),
+                ValueDeletable::Value(Bytes::from_static(b"v2")),
+                9,
+                Some(10000 - 3000),
+                Some(10000 + 4000),
+            )), // not expired
+        ],
+        expected: None,
+    })]
+    #[case(LevelGetExpireTestCase {
+        entries: vec![
+            Some(RowEntry::new(
+                Bytes::from_static(b"key"),
+                ValueDeletable::Value(Bytes::from_static(b"v1")),
+                10,
+                Some(10000 - 2000),
+                Some(10000 + 1000),
+            )), // not expired
+            Some(RowEntry::new(
+                Bytes::from_static(b"key"),
+                ValueDeletable::Value(Bytes::from_static(b"v2")),
+                9,
+                Some(10000 - 5000),
+                Some(10000 - 4000),
+            )), // expired
+        ],
+        expected: Some(Bytes::from_static(b"v1")),
+    })]
+    #[case(LevelGetExpireTestCase {
+        entries: vec![
+            None,
+            Some(RowEntry::new(
+                Bytes::from_static(b"key"),
+                ValueDeletable::Value(Bytes::from_static(b"v2")),
+                9,
+                Some(10000 - 5000),
+                Some(10000 + 4000),
+            )), // not expired
+        ],
+        expected: Some(Bytes::from_static(b"v2")),
+    })]
+    #[case(LevelGetExpireTestCase {
+        entries: vec![
+            None,
+            Some(RowEntry::new(
+                Bytes::from_static(b"key"),
+                ValueDeletable::Tombstone,
+                9,
+                Some(10000 - 5000),
+                Some(10000 + 4000),
+            )), // tombstone
+            Some(RowEntry::new(
+                Bytes::from_static(b"key"),
+                ValueDeletable::Value(Bytes::from_static(b"v2")),
+                9,
+                Some(10000 - 5000),
+                None, // no expiration
+            )), // not expired
+        ],
+        expected: None,
+    })]
+    #[case(LevelGetExpireTestCase {
```

Comment:
this is an exact copy of the test case above it I think?

Reply:
oh, there's a nit difference: this test case does not started with a `None` as the first element.

Reply:
ah, subtle. thanks 👍